head/src/main/kotlin/io/qalipsis/core/head/report/ReportFileBuilder.kt

```diff
@@ -0,0 +1,127 @@
+package io.qalipsis.core.head.report
+
+import io.aerisconsulting.catadioptre.KTestable
+import io.qalipsis.api.report.TimeSeriesAggregationResult
+import io.qalipsis.api.report.TimeSeriesRecord
+import io.qalipsis.core.head.configuration.HeadConfiguration
+import io.qalipsis.core.head.jdbc.entity.ReportDataComponentEntity
+import io.qalipsis.core.head.jdbc.entity.ReportEntity
+import io.qalipsis.core.head.model.DataComponentType
+import io.qalipsis.core.head.model.Zone
+import jakarta.inject.Singleton
+import java.time.Instant
+
+@Singleton
+internal class ReportFileBuilder(
+    private val timeSeriesDataQueryService: TimeSeriesDataQueryService,
+    private val headConfiguration: HeadConfiguration,
+) {
+    /**
+     * Assemble other parts of the data needed to construct a report.
+     *
+     * @param report instance of the [ReportEntity] from which to populate report file.
+     * @param tenant reference of the tenant owning the report
+     * @param campaignData data class which contains all report data
+     */
+    suspend fun execute(
+        report: ReportEntity,
+        tenant: String,
+        campaignData: Collection<CampaignData>
+    ): CampaignReportDetail {
+        campaignData.forEach { campaign ->
+            val zones = mutableSetOf<Zone>()
+            headConfiguration.zones.forEach { zone ->
+                if (zone.key in campaign.zones) {
+                    zones.add(zone)
+                }
+            }
+            campaign.apply { resolvedZones = zones }
+        }
+        val tableData = mutableListOf<Collection<TimeSeriesRecord>?>()
```

Comment:
Use empty collections instead of null ones to keep your code safe.
```suggestion
        val tableData = mutableListOf<Collection<TimeSeriesRecord>>()
```